packages/hash/frontend/src/pages/_app.page.tsx

```diff
@@ -131,18 +128,112 @@ const App: FunctionComponent<AppProps> = ({
         };
       `}
       />
-    </ApolloProvider>
+    </>
   );
 };
 
-const AppWithTypeSystemContextProvider: FunctionComponent<AppProps> = (
+// Based on the definition of `NextPage`
+export type AppPage<P = {}, IP = P> = NextComponentType<AppContext, IP, P>;
+
+const AppWithTypeSystemContextProvider: AppPage<AppProps, AppInitialProps> = (
   props,
-) => (
-  <TypeSystemContextProvider>
-    <SessionProvider>
-      <App {...props} />
-    </SessionProvider>
-  </TypeSystemContextProvider>
-);
+) => {
+  const { initialAuthenticatedUser } = props;
+
+  return (
+    <TypeSystemContextProvider>
+      <ApolloProvider client={apolloClient}>
+        <AuthInfoProvider initialAuthenticatedUser={initialAuthenticatedUser}>
+          <App {...props} />
+        </AuthInfoProvider>
+      </ApolloProvider>
+    </TypeSystemContextProvider>
+  );
+};
+
+const redirectInGetInitialProps = (params: {
+  appContext: AppContext;
+  location: string;
+}) => {
+  const {
+    appContext: {
+      ctx: { req, res },
+      router,
+    },
+    location,
+  } = params;
+
+  if (req && res) {
+    if (!res.writeHead) {
+      // `res.writeHead` is not a function whilst building next, so return instead.
+      return;
+    }
+    res.writeHead(307, { Location: location });
+    res.end();
+  } else {
+    void router.push(location);
+  }
+};
+
+// The list of page pathnames that should be accessible whether or not the user is authenticated
+const publiclyAccessiblePagePathnames = [
+  "/[account-slug]/[page-slug]",
+  "/login",
+  "/signup",
+];
+
+AppWithTypeSystemContextProvider.getInitialProps = async (appContext) => {
```

Comment:
Same for this, we could move `App`'s getInitialProps function, or just some parts of it to a separate file to make this file more readable, it looks pretty crowded right now with 200+ lines, hard to follow.